js_modules/dagster-ui/packages/ui-core/src/performance/TraceContext.tsx

```diff
@@ -0,0 +1,95 @@
+import {QueryResult} from '@apollo/client';
+import {ReactNode, createContext, useContext, useLayoutEffect, useMemo} from 'react';
+
+export enum CompletionType {
+  SUCCESS = 1,
+  ERROR = 2,
+  CANCELLED = 3,
+}
+
+type TraceContextType = {
+  createDependency: (_name: string) => Dependency | null;
+  addDependency: (_dep: Dependency | null) => void;
+  completeDependency: (_dep: Dependency | null, type: CompletionType) => void;
+};
+
+export const TraceContext = createContext<TraceContextType>({
+  createDependency: (_name: string) => null,
+  addDependency: (_dep) => {},
+  completeDependency: (_dep, _type) => {},
+});
+
+/**
+ * Use this to wrap child react components who should not count
+ * toward display done. Eg. If you're re-using a component that
+ * adds dependencies but you don't want that component or its dependencies
+ * as your own dependency
+ */
+export const OrphanDependenciesTraceContext = ({children}: {children: ReactNode}) => {
+  return (
+    <TraceContext.Provider
+      value={useMemo(
+        () => ({
+          createDependency: () => null,
+          addDependency: () => {},
+          completeDependency: () => {},
+        }),
+        [],
+      )}
+    >
+      {children}
+    </TraceContext.Provider>
+  );
+};
+
+export class Dependency {
+  public readonly name: string;
+
+  constructor(name: string) {
+    this.name = name;
+  }
+}
+
```

Comment:
I imagine in the future we may want to attach metadata to these dependency segments so I made this a class to allow for that extension in the future